File: java/modules/basic/src/main/java/io/v6d/modules/basic/dataframe/DataFrameBuilder.java
/** Copyright 2020-2023 Alibaba Group Holding Limited.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
package io.v6d.modules.basic.arrow;

import static java.util.Objects.requireNonNull;

import io.v6d.core.client.Client;
import io.v6d.core.client.IPCClient;
import io.v6d.core.client.ds.ObjectBase;
import io.v6d.core.client.ds.ObjectBuilder;
import io.v6d.core.client.ds.ObjectMeta;
import io.v6d.core.common.util.VineyardException;
import java.util.HashMap;
import lombok.*;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public class DataFrameBuilder implements ObjectBuilder {
    private Logger logger = LoggerFactory.getLogger(DataFrameBuilder.class);

    private final HashMap<String, ObjectBase> columns;

    public DataFrameBuilder(final IPCClient client) {
      this.columns = new HashMap<String, ObjectBase>();
    }

    public void addColumn(String name, Tensor column) {
      this.columns.put(name, column);
    }

    public void addColumn(String name, TensorBuilder builder) {
      this.columns.put(name, builder);
    }

    public HashMap<String, ObjectBase> getColumns() {
      return this.columns;
    }

    public int getColumnSize() {
      return this.columns.size();
    }

    @Override
    public void build(Client client) throws VineyardException {}

    @Override
    public ObjectMeta seal(Client client) throws VineyardException {
        this.build(client);

        val meta = ObjectMeta.empty();

        meta.setTypename("vineyard::DataFrame");
        meta.setValue("nbytes", 0); // FIXME
        meta.setValue("partition_index_row_", -1); // FIXME
        meta.setValue("partition_index_column_", -1); //FIXME
        meta.setValue("row_batch_index_", 0);

        meta.setValue("__values_-size", columns.size());
        int index = 0;
        for(String key : columns.keySet()) {
          meta.addValue("__values_-key-" + index, key);
          meta.addMember("__values_-value-" + index, columns.get(key).seal(client));
          index ++;
        }
        for(int index = 0; index < columns.size(); ++index) {
          // Seal Column Names.
        }
        return client.createMetaData(meta);
    }
}

*/

File: java/spark/src/main/scala/io/v6d/spark/rdd/DataFrameBuilder.scala
/** Copyright 2020-2023 Alibaba Group Holding Limited.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package io.v6d.spark.rdd

import io.v6d.core.client.{Client, IPCClient}
import io.v6d.core.client.ds.{ObjectBuilder, ObjectFactory, ObjectMeta}
import io.v6d.core.common.util.{ObjectID, VineyardException}
import io.v6d.modules.basic.arrow.{
  Arrow,
  RecordBatchBuilder,
  Schema,
  SchemaBuilder
}
import org.apache.spark.sql.vineyard.DataContext

import org.apache.spark.{OneToOneDependency, Partition, TaskContext}
import org.apache.spark.rdd.RDD
import org.apache.spark.sql.catalyst.InternalRow
import org.apache.spark.sql.{DataFrame, SparkSession}
import org.apache.spark.sql.vectorized.{
  ArrowColumnVector,
  ColumnVector,
  ColumnarBatch
}
import org.apache.spark.sql.types._
import org.apache.spark.sql.internal.SQLConf
import org.apache.arrow.vector.VectorSchemaRoot

import scala.collection.JavaConverters._

/** Provides a Builder to put spark.sql.DataFrame int vineyard as
  * Vineyard::Table.
  */
class DataFrameBuilder(
    private val client: IPCClient,
    private val sparkDF: DataFrame
) extends ObjectBuilder() {

  @throws(classOf[VineyardException])
  override def build(client: Client): Unit = {}

  // scalastyle:off method.length
  @throws(classOf[VineyardException])
  override def seal(client: Client): ObjectMeta = {
    this.build(client);
    val meta = ObjectMeta.empty();
    val timeZoneId = SQLConf.get.sessionLocalTimeZone
    val schema = sparkDF.schema
    val arrowSchema = DataContext.toArrowSchema(schema, timeZoneId)
    val schemaBuilder = SchemaBuilder.fromSchema(arrowSchema)
    val sock = client.getIPCSocket()
    val batches: Array[ObjectID] = sparkDF.rdd
      .mapPartitions(iterator => {
        val localArray = iterator.toArray
        val localClient = new IPCClient(sock)
        val localArrowSchema = DataContext.toArrowSchema(schema, timeZoneId)
        val recordBatchBuilder = new RecordBatchBuilder(
          localClient,
          localArrowSchema,
          localArray.length
        );
        recordBatchBuilder.finishSchema(localClient);
        localArray.zipWithIndex.foreach { case (row, rowId) =>
          schema.toList.zipWithIndex.foreach { case (field, fid) =>
            val builder = recordBatchBuilder.getColumnBuilder(fid)
            if (field.dataType.isInstanceOf[BooleanType]) {
              builder.setBoolean(rowId.toInt, row.getBoolean(fid))
            } else if (field.dataType.isInstanceOf[IntegerType]) {
              builder.setInt(rowId.toInt, row.getInt(fid))
            } else if (field.dataType.isInstanceOf[LongType]) {
              builder.setLong(rowId.toInt, row.getLong(fid))
            } else if (field.dataType.isInstanceOf[FloatType]) {
              builder.setFloat(rowId.toInt, row.getFloat(fid))
            } else if (field.dataType.isInstanceOf[DoubleType]) {
              builder.setDouble(rowId.toInt, row.getDouble(fid))
            } else {
              throw new Exception(
                "Columnar builder for type " + field.dataType + " is not supported"
              )
            }
          }
        }
        val batchMeta = recordBatchBuilder.seal(localClient)
        val id = batchMeta.getId
        Iterable(id).iterator
      })
      .collect()
    meta.setTypename("vineyard::Table")
    meta.setValue("batch_num_", batches.length)
    meta.setValue("num_rows_", -1)
    meta.setValue("num_columns_", sparkDF.schema.size)
    meta.addMember("schema_", schemaBuilder.seal(client))
    meta.setGlobal()
    meta.setValue("partitions_-size", batches.length);
    for ((batch, i) <- batches.zipWithIndex) {
      meta.addMember("partitions_-" + i, batch)
    }
    meta.setNBytes(0)
    client.createMetaData(meta)
  }
  // scalastyle:on method.length
}
